### ISSUE TITLE
cmake should check ffmpeg versions 

### ISSUE THREAD

**Issue by oleg-alexandrov** [LibVNC/libvncserver#366]
I tried to build the latest libvncserver from github. The cmake tool runs just fine, but then the make command errors out when it comes to ffmpeg. This was noticed first time in https://github.com/LibVNC/libvncserver/issues/231, but even though that issue is closed, the problem is still there.

The cmake logic needs to check if the right ffmpeg exists, and if not, it must give a clear error message about the wrong version, rather than dying later during compilation with a very verbose and unhelpful message. 

I finally build from scratch ffmpeg-4, but then the libvncserver cmake script could not find it, and had no cmake variable which allows you to set it. I finally dug that one has to set the environmental variable FFMPEG_DIR to search for the right ffmpeg. That's a hackish way of doing things. 

Even then, it did not set the include directory and some of the libraries right, and was still going to the system.

After I fixed that, other errors showed up. It would not link to the math library by default, so -lm, it would not link to -lzma, -lswresample, etc, etc. 

I know that these quirks come from ffmpeg, yet, since libvncserver depends on it, it should handle this.

In short, libvncserver assumes that you already have all dependencies, and they are already in the right place in the system, or otherwise one runs into a never-ending chase to make things work. 

I ran into this on Ubuntu 16.

**Comment by bk138** [comment]
thank you for your colorful report. to improve the situation, let's work together: which ones are (from your experience) working ffmpeg versions, which ones fail?

**Comment by oleg-alexandrov** [comment]
Thank you for redirecting my "colorful" report to a productive venue. 

My system has ffmpeg version 2.8.15. 

Regretfully, there seems to be no easy way of checking for the ffmpeg version, except by doing 

  ffmpeg -version

when it gives:

ffmpeg version 2.8.15-0ubuntu0.16.04.1
  libavutil      54. 31.100 / 54. 31.100
  libavcodec     56. 60.100 / 56. 60.100
  libavformat    56. 40.101 / 56. 40.101
  libavdevice    56.  4.100 / 56.  4.100
  libavfilter     5. 40.101 /  5. 40.101
  libavresample   2.  1.  0 /  2.  1.  0
  libswscale      3.  1.101 /  3.  1.101
  libswresample   1.  2.101 /  1.  2.101
  libpostproc    53.  3.100 / 53.  3.100

In your case, only the libraries are important and the header files, and not the ffmpeg executable, but regretfully there is no good way to correlate the libraries version to the ffmpeg version. 

I guess one would have to download successive versions of ffmpeg and some time tabulating the library versions and deciding which one is good. 

I also dug through your FindFFMPEG.cmake. It does not set any version, and even though the parent asks for version 3.1.0, that is not enforced. 

In short, some ungrateful work is needed to make this robust. Hopefully users can dig through these issues and at least get a clue for the problem.

**Comment by bk138** [comment]
Maybe https://github.com/Kitware/VTK/blob/master/CMake/FindFFMPEG.cmake ?
